Template version 1.0.0 (#3)
agile-learning-institute/mentorHub-encounter-api: README.md

# template-python-flask-api

# Remaining TODO work
- [ ] [encounter Service](./src/services/encounter_services.py)
- [ ] [MongoIO](./src/utils/mongo_io.py)
- [ ] [server.py](/src/server.py)
- [ ] [All Test Files](./test/)

Now you can remove these instructions from the readme, 

## Overview

> This is a template repo for a simple Flask API that provides Get/Post/Patch services for documents in the encounter encounter. This API uses data from a [backing Mongo Database](https://github.com/agile-learning-institute/mentorHub-mongodb), and sup8090s a [Single Page Application.](https://github.com/agile-learning-institute/mentorHub-encounter-ui)

Reviewer: ```suggestion
This is a template repo for a simple Flask API that provides Get/Post/Patch services for documents in the encounter collection. This API uses data from a [backing Mongo Database](https://github.com/agile-learning-institute/mentorHub-mongodb), and supports a [Single Page Application.](https://github.com/agile-learning-institute/mentorHub-encounter-ui)
```

```bash
pipenv run stepci
```

## Run StepCI load testing
NOTE: Assumes the API is running at localhost:8088

```bash
pipenv run load
```

# Project Layout
- ``/src`` this folder contains all source code
- ``/src/server.py`` is the main entrypoint, which initializes the configuration and registers routes with Flask
- ``/src/config/config.py`` is the singleton config object that manages configuration values and acts as a cache for enumerators and other low volatility data values.
- ``/src/models`` contains helpers related to creating transactional data objects such as breadcrumbs or RBAC tokens
- ``/src/routes`` contains Flask http request/response handlers
- ``/src/services`` service interface that wraps database calls with RBAC, encode/decode, and other business logic
- ``/src/utils/mongo_io.py`` is a singleton that manages the mongodb connection, and provides database io functions to the service layer. 
- ``/test`` this folder contains unit testing, and testing artifacts. The sub-folder structure mimics the ``/src`` folder

# API Testing with CURL

If you want to do more manual testing, here are the curl commands to use

### Test Health Endpoint

> This endpoint sup8090s the Prometheus monitoring standards for a health check endpoint

Reviewer: ```suggestion
This endpoint supports the Prometheus monitoring standards for a health check endpoint
```

```bash
curl http://localhost:8088/api/health/
```

### Test Config Endpoint

```bash
curl http://localhost:8088/api/config/
```

### Test get a Encounter document

```bash
curl http://localhost:8088/api/encounter/eeee00000000000000000001/
```

> ### Test add a Encounter Document 

Reviewer: ```suggestion
### Test add an Encounter Document 
```

```bash
curl -X POST localhost:8088/api/encounter/ \
-H "Content-Type: application/json" \
-d '{"date":"2024-12-07 00:00:00","mentorId":"aaaa00000000000000000005","personId":"aaaa00000000000000000024","planId":"eeff00000000000000000004","status":"Active"}'
```

> ### Test update a Encounter Document

Reviewer: ```suggestion
### Test update an Encounter Document
```

-H "Content-Type: application/json" \
-d '{"status":"Archived"}'
```

### Test get a Plan document

```bash
curl http://localhost:8088/api/plan/eeff00000000000000000002/
```

### Test add a Plan Document 

```bash
curl -X POST localhost:8088/api/plan/ \
-H "Content-Type: application/json" \
-d '{"name":"test"}'
```

### Test update a Plan Document

```bash
curl -X PATCH localhost:8088/api/plan/eeff00000000000000000002 \
-H "Content-Type: application/json" \
-d '{"status":"Archived"}'
```

### Test get a list of People

```bash
curl http://localhost:8088/api/people/
```

### Test get a list of Mentors

```bash
curl http://localhost:8088/api/mentors/
```

## Observability and Configuration

The ```api/config/``` endpoint will return a list of configuration values. These values are either "defaults" or loaded from a singleton configuration file, or an Environment Variable of the same name. Configuration files take precedence over environment variables. The environment variable "CONFIG_FOLDER" will change the location of configuration files from the default of ```./```

The ```api/health/``` endpoint is a [Prometheus](https://prometheus.io) Health check endpoint.

> The [Dockerfile](./Dockerfile) uses a 2-stage build, and sup8090s both amd64 and arm64 architectures. 

Reviewer: ```suggestion
The [Dockerfile](./Dockerfile) uses a 2-stage build, and supports both amd64 and arm64 architectures. 
```
